algorithms/finetune/iql.py

```diff
@@ -0,0 +1,710 @@
+    actor_optimizer = torch.optim.Adam(actor.parameters(), lr=3e-4)
+
+    kwargs = {
+        "max_action": max_action,
+        "actor": actor,
+        "actor_optimizer": actor_optimizer,
+        "q_network": q_network,
+        "q_optimizer": q_optimizer,
+        "v_network": v_network,
+        "v_optimizer": v_optimizer,
+        "discount": config.discount,
+        "tau": config.tau,
+        "device": config.device,
+        # IQL
+        "beta": config.beta,
+        "iql_tau": config.iql_tau,
+        "max_steps": config.offline_iterations,
+    }
+
+    print("---------------------------------------")
+    print(f"Training IQL, Env: {config.env}, Seed: {seed}")
+    print("---------------------------------------")
+
+    # Initialize actor
+    trainer = ImplicitQLearning(**kwargs)
+
+    if config.load_model != "":
+        policy_file = Path(config.load_model)
+        trainer.load_state_dict(torch.load(policy_file))
+        actor = trainer.actor
+
+    wandb_init(asdict(config))
+
+    evaluations = []
+
+    state, done = env.reset(), False
+    episode_step = 0
+    print("Offline pretraining")
+    for t in range(int(config.offline_iterations) + int(config.online_iterations)):
+        if t == config.offline_iterations:
+            print("Online tuning")
+        if t >= config.offline_iterations:
+            episode_step += 1
+            action = actor(
+                torch.tensor(
+                    state.reshape(1, -1), device=config.device, dtype=torch.float32
+                )
+            )
+            if not config.iql_deterministic:
+                action = action.sample()
+            else:
+                noise = (torch.randn_like(action) * config.expl_noise).clamp(
+                    -config.noise_clip, config.noise_clip
+                )
+                action += noise
+            action = torch.clamp(max_action * action, -max_action, max_action)
+            action = action.cpu().data.numpy().flatten()
+            next_state, reward, done, _ = env.step(action)
+
+            real_done = False
+            if done and episode_step < max_steps:
+                real_done = True
+
+            if config.normalize_reward:
+                reward = modify_reward_online(reward, config.env, **reward_mod_dict)
+
+            replay_buffer.add_transition(state, action, reward, next_state, real_done)
+            state = next_state
+            if done:
+                state, done = env.reset(), False
+                episode_step = 0
+
+        batch = replay_buffer.sample(config.batch_size)
+        batch = [b.to(config.device) for b in batch]
+        log_dict = trainer.train(batch)
+        log_dict["offline_iter" if t < config.offline_iterations else "online_iter"] = (
+            t if t < config.offline_iterations else t - config.offline_iterations
+        )
+        wandb.log(log_dict, step=trainer.total_it)
+        # Evaluate episode
```

Comment:
We need to both log training and evaluation episodes. Reporting learning curves on the training episodes is actually more preferable